src/overlay/Peer.cpp

```diff
@@ -861,83 +865,101 @@ getOutboundFloodTimer(Application& app, StellarMessage const& msg)
+        // messages for the same slot and validator against the latest SCP
+        // message and drop
+        bool valueReplaced = false;
+        auto minSlotToRemember = mApp.getHerder().getMinLedgerSeqToRemember();
+
+        auto latest = queue.back();
+        bool valueReplaced = false;
+
+        for (auto it = queue.begin(); it != queue.end();)
         {
-            ++it;
+            if (it->mMessage.envelope().statement.slotIndex < minSlotToRemember)
+            {
+                it = queue.erase(it);
+                dropped++;
+            }
+            else if (mApp.getHerder().shouldReplaceSt(
+                         it->mMessage.envelope().statement,
+                         latest.mMessage.envelope().statement))
+            {
+                if (!valueReplaced)
+                {
+                    // std::swap does not invalidate iterators
+                    // old element placed at the end of the queue, and will be
+                    // erased at the end of the loop
+                    std::swap(queue.back(), *it);
+                    valueReplaced = true;
+                    ++it;
+                }
+                else
+                {
+                    it = queue.erase(it);
+                    dropped++;
+                }
+            }
+            else
+            {
+                ++it;
+            }
         }
     }
-    if (dropped)
+
+    if (dropped && Logging::logTrace("Overlay"))
     {
-        CLOG_TRACE(Overlay, "Dropped {} messages to peer {}", dropped,
+        CLOG_TRACE(Overlay, "Dropped {} {} messages to peer {}", dropped,
+                   (lastEnqueuedMessageType == TRANSACTION ? "tx" : "SCP"),
                    mApp.getConfig().toShortString(getPeerID()));
     }
 }
 
 void
-Peer::maybeShedLoadAndSendNextBatch()
+Peer::maybeShedLoadAndSendNextBatch(
+    std::optional<MessageType> lastEnqueuedMessageType)
 {
     releaseAssert(mRemoteOverlayVersion >=
                   Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL);
     releaseAssert(mRemotePeerWantsFlowControl && *mRemotePeerWantsFlowControl);
 
     // First see if we can drop some messages
-    maybeTrimQueue();
+    if (lastEnqueuedMessageType)
+    {
+        maybeTrimQueue(*lastEnqueuedMessageType);
+    }
 
     auto oldOutboundCapacity = mOutboundCapacity;
-    auto it = mFlowControlledQueue.begin();
-    while (it != mFlowControlledQueue.end() && mOutboundCapacity > 0)
+    for (int i = 0; i < mOutboundQueues.size(); i++)
     {
-        sendAuthenticatedMessage(it->mMessage);
-        getOutboundFloodTimer(mApp, it->mMessage)
-            .Update(mApp.getClock().now() - it->mTimeEmplaced);
-
-        // Only count flood messages towards capacity
-        if (mApp.getOverlayManager().isFloodMessage(it->mMessage))
+        auto& queue = mOutboundQueues[i];
+        auto it = queue.begin();
+        while (it != queue.end() && mOutboundCapacity > 0)
```

Comment:
the code would look better if you use empty/front/pop_front instead of iterator (makes it obvious that we're sending the first message in the queue)